adds insightsv3 api support (basic)

scotwells opened the pull request:
Adds basic support of the insights v3 API. Basic usage below.

``` go
query := NewInsightsV3Query()
query.Dimension1 = "action_agent"
query.Fields = []string{"case_reopens", "resolved_cases", "inbound_interactions"}
// view 10 week period
query.Time.Min = Timestamp{time.Now().Add(-10 * 7 * 24 * time.Hour).UTC()}
query.Time.Max = Timestamp{time.Now().UTC()}
query.Time.WindowSize = "hour"

report, resp, err := client.InsightsV3.Report(query)
```

The `InsightsV3Query` struct is used to provide an easy interface to the [reports](http://dev.desk.com/API/insights3/#reports-create) query syntax. The `InsightsV3Report` struct has a function `GetData` that can be used to retrieve the data as a map where the key is the values corresponding key returned from `header`.

Commit:
adds insightsv3 api support

Insights service can be accessed from client and used to generate
a report from the Desk API

talbright commented:
Sweet...still got more code to contribute, the mans on :fire: :fire: 

I'll check it out hopefully in the AM tomorrow but if not by the end of the week latest.

talbright commented on types/insights_v3_query.go:
```diff
@@ -0,0 +1,38 @@
+package types
```
Good call putting this in types.

talbright commented on service/insights_v3.go:
```diff
@@ -0,0 +1,24 @@
+	client *Client
+}
+
+func (s *InsightsV3Service) Report(query *InsightsV3Query) (*InsightsV3Report, *http.Response, error) {
```
Question: how does this method handle insightsv3 rate limits as described in the docs (http://dev.desk.com/API/insights3/#meta-show)? Note that the rate limiting doesn't utilize header fields like rate limiting does in the rest of the API. 

> Rate Limits
> In addition to the general API rate limits described here, the /v2/insights3/ endpoints have an additional rate limit: due to the long-running nature of these requests and the amount of data that they return, only a single /V2/insights3/ outstanding request is allowed at a time. Any request made while another /v2/insights3/ request is outstanding will receive this body:

``` json
{
  "message": "Too Many Requests"
}
```

I think it will handle the error gracefully but just wanted to be sure you checked. 

Perhaps adding an integration test for this would be a good idea...

scotwells replied:
As long as a 429 response is returned from the API then the changes in #13 should be able to handle it.. but yes an integration test for this is a great idea.

Looks like `X-Rate-Limit-Remaining` is passed along with these requests. Based on #13, if a 429 is returned then the client would sleep until `X-Rate-Limit-Reset` has passed and then the request would be retried. Assuming no other insights requests are being made then the subsequent request should succeed.

talbright replied:
Would be good to check if the 429 response code is used. The headers may be passed but not actually utilized. I only say this after reading the docs (http://dev.desk.com/API/insights3/#meta-show) which granted could be wrong or outdated:

> The general V2 response headers described here (http://dev.desk.com/API/using-the-api/#rate-limits) are not modified by or used to determine the Insights3 concurrency limit.

scotwells replied:
Even if not utilized by the desk API, the client would still follow the `X-Rate-Limit-Reset` after receiving a 429 even if it has no effect on concurrency limit

talbright replied:
To get a better idea of how this is working in the API I wrote a quick script to exercise it: 

https://gist.github.com/talbright/48ec6fa8522b7fa4ce32

So basically yes the insightsv3 endpoint will set the response code to 429 when the concurrency limit is hit.

Secondly, even though the insightsv3 API endpoint isn't setting the `X-Rate-Limit-Reset`, it's still being sent back (base controller or filtering is doing that in the server side).

So in the end, your work in #13 will trigger and delay for `X-Rate-Limit-Reset` seconds and retry. Even though thats not directly related to the concurrency limits of insightsv3 endpoints, it still provides good error handling with retries.

:+1:

talbright replied:
Probably not critical to write an integration test for this, since #13 covers it, but would still be good to get a unit test for `report.go` mentioned below.

talbright commented on resource/report.go:
```diff
@@ -0,0 +1,28 @@
+	Data    [][]interface{}  `json:"data"`
+}
+
+func (r *InsightsV3Report) GetData() []map[string] interface{} {
```
Can you add a unit test for this (`report_test.go`)? Thus far I've resisted against writing unit tests that simply exercise the struct json mappings (though maybe it would have caught #16 that you fixed). But for cases where a resource does provide utility functions, such as https://github.com/talbright/go-desk/blob/60537dd851062dd90102e9c4adc6c19a7b38c8df/resource/customer_test.go, I think it's a good idea to provide test coverage.

talbright commented:
Nice addition! If you could add a little more test coverage that would be perfect, and then I can merge this in.

scotwells commented:
I'm sure I'll have find some more things to add :) 

I thought about creating a separate struct so that the dimension values and window time would be separate from the data fields. This would make it easier to iterate through the data fields without having to worry about ignoring the dimension values and the window time. Any interest in going that route?

``` go
type InsightsV3Response struct {
    Request *InsightsV3Query    `json:"request"`
    Header  []string            `json:"header"`
    Data    []*InsightsV3Report `json:"data"`
}  

type InsightsV3Report struct {
    Dimension1Value interface{}
    Dimension2Value interface{}
    WindowTime      Timestamp
    Data            map[string] interface{}
}
```

I should be able to get the unit tests done sometime today

talbright commented:
That does seem like an improvement, as it would make it easier to navigate the data rows in the report. I would probably keep the top level struct named `InsightsV3Report` and name the data container struct  `InsightsV3ReportDataRow`, `InsightsV3ReportDatum`, `InsightsV3ReportDataItem` or something like that.

Commit:
modify reports api

The report API now allows access to an array of `InsightsV3ReportItem`
that provides easy access to the data that was retrieved from the API.
The `InsightsV3ReportItem` will separates the dimension values and
window time from the requested fields, allowing easier traversal of the
field values.

Commit:
add unit tests

Commit:
add ability to retrieve the groups that a user belongs to

talbright commented:
@scotwells were you still planning on adding those tests?

scotwells commented:
Yeah, still working on them. Trying to find some spare time to crank them out. Almost there though 👍🏻

scotwells commented:
Is there any interest in merging this in? Desk's API documentation is actually incorrect when it comes to this feature so I don't really know everything this is able to do. I had to inspect the requests being sent out by the Business Insights dashboard in order to get it to work properly. I've got tests covering a good amount of the changes, just don't want to spend the extra time if its not worth it :)

talbright commented:
The only thing holding me back from merging was/are the missing tests. I'm not of the mindset that something has to be all or nothing, and this seems like a step in the right direction. 

Do you know where the [documentation](http://dev.desk.com/API/insights3/#meta-show) is wrong and what is missing? I can try and get someone to fix it. I didn't work on this feature, but I know the dev that did so I can ask him. Also if its not a documented API endpoint, its subject to change so we wouldn't want to add it to go-desk.

scotwells commented:
So the documentation for report filters gives this filter structure 

``` json
{ 
  "filters": [
    [
      [
        "Include",
        "Labels",
        [
          {
            "id": "1"
          }
        ]
      ]
    ]
  ]
}
```

But the business insights dashboard is using the following

``` json
{
   "filters":[
      [
         [
            "Include",
            "Agents",
            "Is",
            "",
            false,
            [
               {
                  "id":"123"
               }
            ]
         ]
      ]
   ]
}
```

talbright commented:
@kristin there seems to be some differences between what the published docs say on this endpoint and what is actually being served
